debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbStreamingChangeEventSource.java

```diff
@@ -262,8 +245,65 @@ else if (oplogStart.getTime() > 0) {
+                    // But this is currently immutable.
+                    rsChangeStream = createChangeStream(client, rsOffsetContext);
+                    cursor = rsChangeStream.cursor();
+                }
             }
         }
+        finally {
+            cursor.close();
+        }
+    }
+
+    private ChangeStreamIterable<BsonDocument> createChangeStream(MongoClient client, ReplicaSetOffsetContext rsOffsetContext) {
+        final List<Bson> pipeline = new ChangeStreamPipelineFactory(rsOffsetContext, taskContext.getConnectorConfig(), taskContext.filters().getConfig()).create();
+
+        final ChangeStreamIterable<BsonDocument> rsChangeStream = client.watch(pipeline, BsonDocument.class);
+        if (taskContext.getCaptureMode().isFullUpdate()) {
+            rsChangeStream.fullDocument(FullDocument.UPDATE_LOOKUP);
+        }
+        if (taskContext.getCaptureMode().isIncludePreImage()) {
+            rsChangeStream.fullDocumentBeforeChange(FullDocumentBeforeChange.WHEN_AVAILABLE);
+        }
+
+        if (rsOffsetContext.lastResumeToken() != null) {
+            LOGGER.info("Resuming streaming from token '{}'", rsOffsetContext.lastResumeToken());
+            rsChangeStream.resumeAfter(ResumeTokens.fromData(rsOffsetContext.lastResumeToken()));
+        }
+        else if (rsOffsetContext.lastOffsetTimestamp().getTime() > 0) {
+            LOGGER.info("Resume token not available, starting streaming from time '{}'", rsOffsetContext.lastOffsetTimestamp());
+            rsChangeStream.startAtOperationTime(rsOffsetContext.lastOffsetTimestamp());
+        }
+
+        if (connectorConfig.getCursorMaxAwaitTime() > 0) {
+            rsChangeStream.maxAwaitTime(connectorConfig.getCursorMaxAwaitTime(), TimeUnit.MILLISECONDS);
+        }
+        return rsChangeStream;
+    }
+
+    private static boolean isSelectedReadPreference(MongoClient client, ReadPreference readPreference,
+                                                    MongoCursor<?> cursor) {
+        // Find all remaining nodes that match our preference
+        var candidates = new ReadPreferenceServerSelector(readPreference)
+                .select(client.getClusterDescription());
+
+        // Determine if the cursor matches any one of these candidates
+        return candidates.stream()
+                .map(ServerDescription::getAddress)
+                .anyMatch(preferredAddress -> {
+                    // TODO: In general, the host names returned between commands can be different (DNS or IP potentially).
+                    // For example, we see test-mongo1:27017 (defined in rs.init()) and e253ef34bd93:27017 (the container’s hostname).
+                    // Thus, canonicalization must be performed when making comparison.
+                    var cursorAddress = cursor.getServerCursor() == null ? null : cursor.getServerCursor().getAddress();
+                    return preferredAddress.equals(cursorAddress);
```

Comment:
This is my biggest concern with the approach as it could lead to false positives in certain configurations / topologies. Not sure if there is a better way unfortunately, but open to ideas.